internal/storage/chunk_uploader_test.go

```diff
@@ -0,0 +1,463 @@
+
+type ChunkUploaderTest struct {
+	fakeStorage FakeStorage
+	content     string
+	req         *gcs.CreateObjectRequest
+	obj         *storage.ObjectHandle
+}
+
+var _ SetUpInterface = &ChunkUploaderTest{}
+var _ TearDownInterface = &ChunkUploaderTest{}
+
+func init() { RegisterTestSuite(&ChunkUploaderTest{}) }
+
+func (t *ChunkUploaderTest) SetUp(ti *TestInfo) {
+	t.fakeStorage = NewFakeStorage()
+	storageHandle := t.fakeStorage.CreateStorageHandle()
+	bucketHandle := storageHandle.BucketHandle(TestBucketName, "")
+
+	AssertNe(nil, bucketHandle)
+
+	t.content = "Creating a new object"
+	t.req = &gcs.CreateObjectRequest{
+		Name:     "test_object",
+		Contents: strings.NewReader(t.content),
+	}
+
+	t.obj = bucketHandle.bucket.Object(t.req.Name)
+
+	AssertNe(nil, t.obj)
+	AssertEq("test_object", t.obj.ObjectName())
+}
+
+func (t *ChunkUploaderTest) TearDown() {
+	t.fakeStorage.ShutDown()
+}
+
+////////////////////////////////////////////////////////////////////////
+// Helper functions
+////////////////////////////////////////////////////////////////////////
+
+// testNewChunkUploader tests a newky created ChunkUploader and asserts
+// if it hasn't been created properly.
+// If properly created, it casts it to chunkUploader pointer and returns.
+func (t *ChunkUploaderTest) testNewChunkUploader(ctx context.Context, obj *storage.ObjectHandle, req *gcs.CreateObjectRequest, chunkSize int, progressFunc func(int64)) *chunkUploader {
+	uploader, err := NewChunkUploader(ctx, obj, req, chunkSize, progressFunc)
+	AssertNe(nil, uploader)
+	AssertEq(nil, err)
+
+	chunkUploader := uploader.(*chunkUploader)
+
+	AssertNe(nil, chunkUploader)
+	AssertEq(Initialized, chunkUploader.state)
+	AssertEq(0, chunkUploader.totalWriteInitiatedSoFar)
+	AssertEq(0, chunkUploader.BytesUploadedSoFar())
+	AssertNe(nil, chunkUploader.writer)
+	AssertEq(req.Name, chunkUploader.objectName)
+	AssertEq(chunkSize, chunkUploader.writer.ChunkSize)
+
+	return chunkUploader
+}
+
+func (t *ChunkUploaderTest) testUploadChunkAsync(uploader *chunkUploader, contents io.Reader) {
+	err := uploader.UploadChunkAsync(contents)
+	AssertEq(nil, err)
+	AssertEq(Uploading, uploader.state)
+}
+
+func (t *ChunkUploaderTest) testUploadEmptyContent(uploader *chunkUploader, contents io.Reader) {
+	err := uploader.UploadChunkAsync(contents)
+	AssertEq(nil, err)
+}
+
+func (t *ChunkUploaderTest) testFailedUploadChunkAsync(uploader *chunkUploader, contents io.Reader) {
+	err := uploader.UploadChunkAsync(contents)
+	AssertNe(nil, err)
+	AssertEq(UploadError, uploader.state)
+}
+
+func (t *ChunkUploaderTest) testClose(uploader *chunkUploader, objName string, len int) {
+	o, err := uploader.Close()
+	AssertEq(nil, err)
+	AssertNe(nil, o)
+	AssertEq(Closed, uploader.state)
+	AssertEq(objName, o.Name)
+	AssertEq(len, o.Size)
+}
+
+func (t *ChunkUploaderTest) testFailedClose(uploader *chunkUploader) {
+	_, err := uploader.Close()
+	AssertNe(nil, err)
```

Comment:
Should we assert on the error string as well?

Reply:
Added assert on error substring.